src/uucore/src/lib/features/i18n/mod.rs

```diff
@@ -0,0 +1,46 @@
+use std::{cmp::Ordering, sync::OnceLock};
+
+use icu::{
+    collator::{Collator, options::CollatorOptions},
+    locale::{Locale, locale},
+};
+
+fn get_collating_locale() -> &'static Locale {
+    static COLLATING_LOCALE: OnceLock<Locale> = OnceLock::new();
+
+    COLLATING_LOCALE.get_or_init(|| {
+        // Look at 3 environment variables in the following order
+        //
+        // 1. LC_ALL
+        // 2. LC_COLLATE
+        // 3. LANG
+        //
+        // Or fallback on Posix locale
+
+        let locale_var = std::env::var("LC_ALL")
+            .or_else(|_| std::env::var("LC_COLLATE"))
+            .or_else(|_| std::env::var("LANG"));
+
+        if let Ok(locale) = locale_var {
+            if let Some(simple) = locale.split(&['-', '@']).next() {
+                let bcp127 = simple.replace("_", "-");
+                if let Ok(locale) = Locale::try_from_str(&bcp127) {
+                    return locale;
+                }
+            }
+        }
+        // Default POSIX locale representing LC_ALL=C
+        locale!("en-US-posix")
+    })
+}
+
+pub fn locale_compare<T: AsRef<str>>(left: T, right: T) -> Ordering {
+    let collator = Collator::try_new(get_collating_locale().into(), CollatorOptions::default());
```

Comment:
It's inefficient to construct a new collator for every comparison. The API design intent is that a `CollatorBorrowed` instance is used across multiple comparisons that use the same configuration. If you want to have a function that just takes the operands to be compared without the caller having to manage keeping around the `CollatorBorrowed`, one option is to declare `static mut COLLATOR: Option<CollatorBorrowed> = None;` and assign `Some(CollatorBorrowed::try_new(...).unwrap())` on process startup before performing any comparisons.

Once a `Some(CollatorBorrowed)` exists, it is thread-safe to do `COLLATOR.as_ref().compare(...)`.